### ISSUE TITLE
InputNode and OutputNode might be DataNodes?

### ISSUE THREAD

**Issue by HRogge** [CZ-NIC/yangson#53]
Hi... I just discovered a few 'strange' things while working on xml/json import/export and my own Restconf server and the RPC implementation. My main issue at the moment is that InputNode and OutputNode are not DataNodes... but the RPC input and output contains the "input/output" tag. While you can get around this with stripping of one layer of JSON, doing the from/to-json/xml part and then re-adding them, making them inhert DataNode seems to be more clean...

but this lead me to a single yangson testcase that goes wrong... the ascii-tree testcase.
at the end you have the 'correct' output:
```
+---x testb:rpcA
   +--ro input
   |  +--ro testb:leafK? <typA(int16)>
   +--ro output
      +--ro testb:llistC* <boolean>
```

is this correct? the rpcA node has the same namespace than the childs of input/output, but you have the explicit namespace.

I noticed this namespace vanishes when you change InputNode/Output node to inhert from DataNode.

What do you think?

**Comment by HRogge** [comment]
I added pull request #54 for this change to make it easier to review the change.

**Comment by kwatsen** [comment]
I don't think this is correct.  I think the "input" and "output" nodes are chameleon, in that they implicitly inherit the namespace from the parent.

BTW, the example above actually *shows* the `input` and `output` nodes as being in the `testb` namespace (i.e., when no namespace is given, the parent's namespace is used).  That `leafK` and `llistC` both declare the same namespace is unnecessary...and potentially wrong, depending on how strictly namespace-compression is observed.

Sadly, [RFC 8340](https://tools.ietf.org/html/rfc8340) doesn't speak to this directly, but both `pyang` and `yanglint` generate tree-diagrams like this (without superfluous prefixes):
```
module: foo

  rpcs:
    +---x bar
    |  +--- input
    |  |  +--- ...
    |  +--- output
    |     +--- ...
```

Warning: `yangson` may be incorrectly prefixing the `rpcA`, since the RPC is already in the module's namespace.

FWIW, RFC 7950 sections [7.14](https://tools.ietf.org/html/rfc7950#section-7.14) and [7.15](https://tools.ietf.org/html/rfc7950#section-7.15) aren't much help, but note how in the examples listed in [Section 3.6 of RFC 8040](https://tools.ietf.org/html/rfc8040#section-3.6) show the "input" and "output" nodes in the parent namespace too.

**Comment by HRogge** [comment]
Maybe my initial message in this issue was not that good...

with my patch the "rpcA" namespace tag of the input/output nodes child vanishes, as you also have done in your diagram. The patch fixes the prefixing of "rpcA".

the second thing this patch is doing is allowing the input/output node inherit from DataNode, which makes parsing the subtrees of rpc input/output much easier.

**Comment by llhotka** [comment]
This is a messy issue because in the encoding defined in RFC 7950 (i.e. for NETCONF) `input` and `output` doesn't appear.

**Comment by llhotka** [comment]
The tree diagram in yangson isn't done module-wise as in pyang, so all top-level nodes need to have a module prefix. Otherwise, how could you tell from the diagram that `rpcA` is defined in `testb`?

**Comment by kwatsen** [comment]
@llhotka  

Right, RFC 7950 isn't clear, but I think we can infer what the `input` and `output` nodes implicitly inherit the namespace from the parent `RPC` or `action.

Also, regarding why prefixing the RPCs is needed, that makes sense, thanks.

**Comment by llhotka** [comment]
I am more worried about making `InputNode` and `OutputNode` into subclasses of `DataNode` because it could potentially break existing code - e.g. the method `InternalNode::get_data_child` could return different results.

**Comment by HRogge** [comment]
I will have a look at the server code tomorrow, maybe I can remember what was the trouble with Input/Output not a Datanode... there might be a different way to solve this.

**Comment by HRogge** [comment]
I did some fixing on my submodules branch and some testing with my server and I think I see the (at least) two issues I had...

the first is easy to solve, using the get_child() instead of get_data_child() function inside from_raw inside the InternalNode class.

the second one looks more difficult to do. I am using "full_peek()" of the Instance class (an extended version of peek() which gives you schema and instance), but this doesn't work correctly for a non-data node in the path. Which means when I try to peek/goto a path like "/ietf-restconf:restconf/operations/example-ops:get-reboot-info", I get an error that the path does not exist.

This (most likely) could also be fixed by introducing a "allow_nodata" optional parameter to peek.

**Comment by kwatsen** [comment]
I'm having an issue that may or may not be related to this `Issue`.

My app's core logic is all Python-object based.  Receiving an JSON-encoded RPC, it works (in my app's constrained use-case) to import the `input` using `json.loads()`.   Now I'm trying to support XML-encoded RPCs, and there is no such thing as`xml.loads()`, so I'm trying to use a combination of Yangson's new `from_xml()` and its existing `raw_value()` methods to do the XML-to-Python-object conversion for me.

But I'm having an issue, that essentially boils down to not knowing how to get Yangson to export a "raw value" after using `from_xml()`.  Related to this `Issue`, I wonder if it might be because `InputNode` is not a `DataNode`...

To be clear, `raw_value()` is **only** a method of the `InstanceNode` class, which can be obtained via the `DataModel` class's `from_xml()` method.  However, in order to import an RPC's `input` (or, I imagine, any `InteriorNode`), the `DataModel.get_schema_node()` method must be used to get the path-specific `SchemaNode`, on which its `from_xml()` method can be used to import the XML string.  However, `SchemaNode.from_xml()` (and `from_raw()`) returns an `InstanceValue` (not an `InstanceNode`) and I can't find anyway to convert it to an `InstanceNode`, and hence no way to call `raw_value()`...

Thoughts?

**Comment by HRogge** [comment]
Thank you @kwatsen , that was the hint I needed to remember more of my troubles...

I will just post the "handle RPC action" code I use in my server, this might show you what is possible if input/output are data nodes:

```
    def action(self, body, json_in: bool, json_out: bool, p_schema):
        rpc = self.rcdata.rpc.get(p_schema.data_path())
        if not rpc:
            raise RestconfError(500, emsg="No handler registered for rpc: " + p_schema.data_path())

        input_schema = p_schema.get_child("input")
        output_schema = p_schema.get_child("output")

        if input_schema.children:
            if not body:
                raise RestconfError(emsg='Input missing for RPC: ' + p_schema.data_path())
            if json_in and len(body) != 1:
                raise RestconfError(emsg='Input with multiple root nodes for RPC: ' + p_schema.data_path())
            try:
                if json_in:
                    input_value = p_schema.from_raw(body, allow_nodata=True)
                else:
                    input_value = p_schema.from_xml(body, allow_nodata=True)
            except RawMemberError as e:
                raise RestconfError(emsg='illegal input for RPC {}: {}'.format(
                    p_schema.data_path(), str(e)))
            except MissingModuleNamespace as e:
                raise RestconfError(emsg='unknown xml namespace: ' + e.ns)
        elif body:
            raise RestconfError(400, emsg='RPC does not expect input: ' + p_schema.data_path())
        else:
            input_value = None

        try:
            output_raw = rpc(self.rcdata, p_schema, input_value)
        except BaseException as e:
            if isinstance(e, RestconfError):
                raise e
            raise RestconfError(500, emsg='An exception happened: ' + str(e))

        if output_schema.children:
            if not output_raw:
                raise RestconfError(500, emsg='RPC did not generate output')

            output_value = p_schema.from_raw(output_raw, allow_nodata=True)
            output_instance = RootNode(output_value, p_schema, self.rcdata.dm.schema_data, output_value.timestamp)
            if json_out:
                result = output_instance.raw_value()
            else:
                firstkey = tuple(output_value.keys())[0]
                result = output_instance[firstkey].to_xml()
        elif output_raw:
            raise RestconfError(500, emsg='RPC does not expect output: ' + str(output_raw))
        else:
            result = None

        response = flask.Response(
            status=200 if result else 204,
            headers={}
        )
        return response, result

```

**body** is the HTTP body of the RPC action, **p_schema** is the parent schema of the RPC (the parent of input/output), **self.rcdata.rpc** is a dictionary that maps a JSON path to a handler function for the RPC.
This code handles both JSON and XML (in theory it could handle different input and output format) and it validates both the input and the output of the RPC against the YANG

**Comment by llhotka** [comment]
BTW, I changed the ASCII tree output to handle namespaces consistently: **any** schema node is now printed with its namespace (module name) if and only if its namespace differs from the parent's, or there is no parent.

**Comment by kwatsen** [comment]
Hi Lada, 

Please see the last couple comments, mine and then Henning's.   There is an issue for which InputNode/OutputNode subclassing DataNode appears to solve.  You wrote [before](https://github.com/CZ-NIC/yangson/issues/53#issuecomment-733811856) about concern for what said change might possibly break.  Could you suggest a path forward?

PS: the use-case in my comment could've been more compellingly-stated as follows:  Imagine an RPC that needs to returns opstate (e.g., statistics).  The app-logic generates a response using Python objects, `dict`, `list`, `str`, etc.  (not Yangson InstanceValue objects).   A Yangson-guided transformation is needed to serialize the response to the desired encoding (e.g., JSON, XML, etc.).   This transformation is some combination of from_raw() and, e.g., to_xml()

Thanks,
Kent

**Comment by llhotka** [comment]
OK, it is merged in version 1.3.56. We'll see if somebody is affected.
